[HLSL] Implement SV_GroupThreadId semantic

lizhengxing opened the pull request:
Support HLSL SV_GroupThreadId attribute.

For `directx` target, translate it into `dx.thread.id.in.group` in clang codeGen and lower `dx.thread.id.in.group` to `dx.op.threadIdInGroup` in LLVM DirectX backend.

For `spir-v` target, translate it into `spv.thread.id.in.group` in clang codeGen and lower `spv.thread.id.in.group` to a `LocalInvocationId` builtin variable in LLVM SPIR-V backend.

Fixes: #70122

Commit:
[HLSL] Implement SV_GroupThreadId semantic

Support SV_GroupThreadId attribute.
Translate it into dx.thread.id.in.group in clang codeGen.

Fixes: #70122

inbelic commented on clang/test/SemaHLSL/Semantics/valid_entry_parameter.hlsl:
```diff
@@ -49,3 +49,28 @@ void CSMain3_GID(uint3 : SV_GroupID) {
+// CHECK-NEXT: HLSLSV_GroupThreadIDAttr
+}
+[numthreads(8,8,1)]
+void CSMain3_GThreadID(uint3 : SV_GroupThreadID) {
```
Out of curiosity, what is the practical use of this? If there isn't why do we accept it?

tex3d replied:
The general use case could be for parameters that contribute to the inter-stage signature.  You must still declare the matching set of parameters in the same order, even if you don't use them, to match signatures between stages.  This doesn't make a difference in this stage (cs), but it would still be a generally supported pattern for all function signatures in HLSL.

inbelic reviewed (commented):
Do we need to consider how spirv handles this? Otherwise, LGTM

lizhengxing commented:
> Do we need to consider how spirv handles this? Otherwise, LGTM

This PR is similar to the one for SV_GroupID. We don't consider the spirv part for both PRs.

llvm-beanz commented:
> This PR is similar to the one for SV_GroupID (#117781). We don't consider the spirv part for both PRs.
I believe that was a mistake, and I don't believe #117781 had sufficient reviews to have merged. It's fine, but I've filed #118700 to complete the outstanding work.

tex3d commented on clang/test/SemaHLSL/Semantics/entry_parameter.hlsl:
```diff
@@ -2,15 +2,18 @@
-// expected-error@+1 {{attribute 'SV_GroupID' is unsupported in 'mesh' shaders, requires compute}}
-void CSMain(int GI : SV_GroupIndex, uint ID : SV_DispatchThreadID, uint GID : SV_GroupID) {
-// CHECK: FunctionDecl 0x{{[0-9a-fA-F]+}} <{{.*}}> line:[[@LINE-1]]:6 CSMain 'void (int, uint, uint)'
+// expected-error@+4 {{attribute 'SV_GroupIndex' is unsupported in 'mesh' shaders, requires compute}}
+// expected-error@+3 {{attribute 'SV_DispatchThreadID' is unsupported in 'mesh' shaders, requires compute}}
+// expected-error@+2 {{attribute 'SV_GroupID' is unsupported in 'mesh' shaders, requires compute}}
+// expected-error@+1 {{attribute 'SV_GroupThreadID' is unsupported in 'mesh' shaders, requires compute}}
+void CSMain(int GI : SV_GroupIndex, uint ID : SV_DispatchThreadID, uint GID : SV_GroupID, uint GThreadID : SV_GroupThreadID) {
```
Why is this using mesh shader as the example of a stage where these are illegal?  These are actually legal for the mesh shader stage.

lizhengxing replied:
@tex3d  Yes, you're right. I checked the spec https://microsoft.github.io/DirectX-Specs/d3d/MeshShader.html .  Those 4 attributes are legal for the mesh shader.

I think that they are legal for the amplification shader as well.

It looks like that we did something wrong in `SemaHLSL::CheckSemanticAnnotation` function
```
  switch (AnnotationAttr->getKind()) {
  case attr::HLSLSV_DispatchThreadID:
  case attr::HLSLSV_GroupIndex:
  case attr::HLSLSV_GroupThreadID:
  case attr::HLSLSV_GroupID:
    if (ST == llvm::Triple::Compute)
      return;
    DiagnoseAttrStageMismatch(AnnotationAttr, ST, {llvm::Triple::Compute});
    break;
```
We only check `Compute Shader` triple. I think that we should add checks for  `Mesh Shader` and `Amplification Shader` triples here. Right?

tex3d replied:
We don't yet support those stages generally, and I don't think this will be the general diagnostic approach going forward anyway, so I'm not sure we need to update this code with those stages at this time.

Perhaps we can simply choose another shader stage in the test for which these are not allowed for now, such as `vertex`.

tex3d replied:
Though you won't be able to use the same entry with the `[numthreads(...)]` attribute for a vertex shader.  Perhaps just add that as a define on the command line?  Or better yet, use a separate test file, since the shared content is extremely minimal between the tests.  That would certainly be less confusing.

lizhengxing replied:
@tex3d It makes sense to me. I'll try to move the non-CS checks into `invalid_entry_parameter.hlsl`

lizhengxing replied:
@tex3d Done. Updated the test in this commit https://github.com/llvm/llvm-project/pull/117781/commits/dc8d779f067e5b8d22e56036c4ba7320e297f339

Commit:
Don't test the Group/Thread input IDs with mesh shader

The SV_GroupIndex, SV_DispatchThreadID, SV_GroupID and SV_GroupThreadID are actually legal for meash shader stage. It shouldn't test them with mesh shader.

This commit tests them with vertex shader and move the test into invalid_entry_parameter.hlsl which's a better place for it.

Commit:
[HLSL][SPIR-V] Add SV_GroupThreadID semantic support

The HLSL SV_GroupThreadID semantic attribute is lowered into @llvm.spv.thread.id.in.group intrinsic in LLVM IR for SPIR-V target.

In the SPIR-V backend, this is now correctly translated to a `LocalInvocationId` builtin variable.

Fixes #70122

lizhengxing commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -389,6 +389,11 @@ llvm::Value *CGHLSLRuntime::emitInputSemantic(IRBuilder<> &B,
   }
+  if (D.hasAttr<HLSLSV_GroupThreadIDAttr>()) {
+    llvm::Function *GroupThreadIDIntrinsic =
+        CGM.getIntrinsic(Intrinsic::dx_thread_id_in_group);
```
@llvm-beanz Done. Added the SPIR-V lowering in this commit https://github.com/llvm/llvm-project/pull/117781/commits/28f823454873d4bc029f2ec57bed3a9707bbf1b2

tex3d commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3606,6 +3613,32 @@ bool SPIRVInstructionSelector::selectSpvThreadId(Register ResVReg,
+bool SPIRVInstructionSelector::selectSpvThreadId(Register ResVReg,
+                                                 const SPIRVType *ResType,
+                                                 MachineInstr &I) const {
+  // DX intrinsic: @llvm.dx.thread.id(i32)
+  // ID  Name      Description
+  // 93  ThreadId  reads the thread ID
+  //
+  // In SPIR-V, llvm.dx.thread.id maps to a `GlobalInvocationId` builtin
+  // variable
```
This comment block seems odd, since we should be translating a SPIRV intrinsic: `@llvm.spv.thread.id`, not a DX intrinsic.  Same for the one in `selectSpvGroupThreadId`.

lizhengxing replied:
@tex3d Done. The comment was updated in the latest commit of the PR.

tex3d commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3606,6 +3613,32 @@ bool SPIRVInstructionSelector::selectSpvThreadId(Register ResVReg,
+  //
+  // In SPIR-V, llvm.dx.thread.id maps to a `GlobalInvocationId` builtin
+  // variable
+  return loadVec3BuiltinInputID(SPIRV::BuiltIn::GlobalInvocationId, ResVReg,
+                                ResType, I);
```
I'm not sure I see any advantage of wrapping these calls in these functions rather than calling `loadVec3BuiltinInputID` directly from the switch case in `selectIntrinsic`.

lizhengxing replied:
@tex3d Yeah, I noticed this as well. I don't call `loadVec3BuiltinInputID` directly just for keep the change minimal.

tex3d replied:
I don't see why we would need to change it in the future, and in any case, adding a function to do something more complicated can always be done in the future if necessary.

lizhengxing replied:
@tex3d  Done. Updated the PR based on your comments.

Commit:
Remove the redundant wrapping code based on Tex's comment

tex3d commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2851,7 +2851,21 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
     break;
   case Intrinsic::spv_thread_id:
-    return selectSpvThreadId(ResVReg, ResType, I);
+    // The HLSL SV_DispatchThreadID semantic is lowered to llvm.spv.thread.id
+    // intrinsic in LLVM IR for SPIR-V backend.
+    //
+    // In SPIR-V backend, llvm.spv.thread.id is now correctly translated to a
+    // `GlobalInvocationId` builtin variable
```
These comments seem redundant to me.  We are in the case handling `llvm.spv.thread.id` and the code translates it to use the `GlobalInvocationId` builtin variable.  What else needs to be said in this context?

"is now correctly translated to ..." as opposed to when?  What does this comment add/clarify?

As far as the HLSL context is concerned, that's defined elsewhere, and it's presumably not in scope for the SPIR-V backend to comment on where an intrinsic may come from in a particular source language.  I can imagine similar expository comments for many other intrinsics being lowered here, but I don't think this is the best place to document such things.

Note: this is just an opinion, feel free to disagree, and I'm curious how others more experienced with the backends feel as well.

lizhengxing replied:
@tex3d Those comments are based on the original comments in `selectSpvThreadId` function with minimal changes. They were added by @sudonatalie. 

If there's no objections from other reviewers, I think it's fine to keep those comments here.

tex3d reviewed (approved):
I think it's good, with minor nit about unnecessary comments.

lizhengxing commented:
@sudonatalie
Hello Natalie,
Could you help review this PR when you're available since it will modify your changes (https://github.com/llvm/llvm-project/pull/82536), Thanks.
